### PR TITLE
feat(eslint-plugin): add new rule prefer-as-const

### LINKED ISSUE #312: Rule proposal: prefer `as const`
In a lot of our code we have to cast string literals to literals types to conform to types so assignments work:

```ts
  type Styles = {
    position: 'relative' | 'absolute'
  }
  
  let styles = {
    position: 'relative' as 'relative'
  }
  
  let s: Styles = styles 
```

This is more common when using React HOCs but above example demonstrate the problem well enough. 

[With TypeScript 3.4 we'll have the `as const` literal expression](https://github.com/Microsoft/TypeScript/pull/29510) which should be preferred to string or number literal types in this case because changing the value does not require changing the type.

Our example code would be better written as:

```ts
type Styles = {
  position: "relative" | "absolute";
};

let styles = {
  position: "relative" as const
};

let s: Styles = styles;
```

**More examples**:

```ts
let a: 'str' = 'str'; // bad 
let a = <'str'>'str'; // bad 
let a = 'str' as 'str' //bad 

let a = 'str' as const // good
```

```ts
// bad
let o = {
    foo: 'string' as 'string'
}

// good
let o = {
    foo: 'string' as const
}
```

### Rule name
Please suggest rule name. I'm not sure how this rule should be named

### Options
I can't think of any configurability for this rule

### PR DESCRIPTION
**Pull request by armano2**
fixes #312

### CODE REVIEW

File: packages/eslint-plugin/src/rules/prefer-as-const.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        valueNode.type === AST_NODE_TYPES.Literal &&
+        typeNode.type === AST_NODE_TYPES.TSLiteralType &&
+        'raw' in typeNode.literal &&
+        valueNode.raw === typeNode.literal.raw
+      ) {
+        context.report({
+          node: typeNode,
+          messageId: 'preferAsConst',
+          fix: canFix
+            ? (fixer): RuleFix => fixer.replaceText(typeNode, 'const')
```
Comment (armano2) [inline packages/eslint-plugin/src/rules/prefer-as-const.ts]: eslint was complaining `explicit-function-return-type`

File: packages/eslint-plugin/docs/rules/prefer-as-const.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Prefer usage of `as const` over literal type (`prefer-as-const`)
+
+This rule recommends usage of `const` assertion when type primitive value is equal to type.
```
Comment (armano2) [inline packages/eslint-plugin/docs/rules/prefer-as-const.md]: I should describe it better but i have no inspiration

File: packages/eslint-plugin/src/rules/prefer-as-const.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+          context.report({
+            node: typeNode,
+            messageId: 'preferConstAssertion',
+            fix: fixer => fixer.replaceText(typeNode, 'const'),
+          });
+        } else {
+          context.report({
+            node: typeNode,
+            messageId: 'variableConstAssertion',
+          });
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/prefer-as-const.ts]: nit - could use a suggestion in this case instead. Suggestions aren't applied automatically, and have to be applied via an IDE, so it allows you to have dangerous and "breaking" fixers.

```ts
          context.report({
            node: typeNode,
            messageId: 'variableConstAssertion',
            suggestions: [
              {
                messageId: 'someId',
                fix: fixer => { ... },
              },
            ],
          });
```

`no-explicit-any` and `no-non-null-assertion` both have fixers (same as #1469)

Comment (armano2) [reply packages/eslint-plugin/src/rules/prefer-as-const.ts]: i'm not familiar with suggestion API, but from what i see its just working same way as fixer but its optional?

**Comment by armano2** [comment]
as recommended i added 2 optional fixers for variable case

File: packages/eslint-plugin/src/rules/prefer-as-const.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        } else {
+          context.report({
+            node: typeNode,
+            messageId: 'variableConstAssertion',
+            suggest: [
+              {
+                messageId: 'variableSuggest1',
+                fix: (fixer): TSESLint.RuleFix =>
+                  fixer.insertTextAfter(valueNode, ' as const'),
+              },
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/prefer-as-const.ts]: how come you have this fixer?
It seems weird to suggest fixing `const x: 2 = 2;` to `const x: 2 = 2 as const;`

Comment (armano2) [reply packages/eslint-plugin/src/rules/prefer-as-const.ts]: ok i removed this suggestion

**Pull request merged**